File: backend/app/schemas/trends/trends.py
"""
Pydantic schemas for Property Trends API.
"""

from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import datetime, date


# ============================================
# Region Schemas
# ============================================

class TrendsRegionBase(BaseModel):
    """Base schema for region data."""
    id: int
    name: str
    name_urdu: Optional[str] = None
    level: int = 2
    
    model_config = ConfigDict(from_attributes=True)


class TrendsRegionResponse(TrendsRegionBase):
    """Response schema for region with timestamps."""
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================
# City Schemas
# ============================================

class TrendsCityBase(BaseModel):
    """Base schema for city data."""
    id: int
    name: str
    name_urdu: Optional[str] = None
    region_id: Optional[int] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    level: int = 3


class TrendsCityResponse(TrendsCityBase):
    """Response schema for city with region info."""
    region: Optional[TrendsRegionBase] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)


class TrendsCityWithLocations(TrendsCityResponse):
    """City response with list of locations."""
    buying_locations: List["TrendsLocationResponse"] = []
    renting_locations: List["TrendsLocationResponse"] = []
    
    model_config = ConfigDict(from_attributes=True)


# ============================================
# Location Schemas
# ============================================

class TrendsLocationBase(BaseModel):
    """Base schema for location data."""
    id: int
    title: str
    title_urdu: Optional[str] = None
    city_id: Optional[int] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    
    model_config = ConfigDict(from_attributes=True)


class TrendsLocationResponse(TrendsLocationBase):
    """Response schema for location."""
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)


class TrendsLocationWithStats(TrendsLocationResponse):
    """Location response with monthly stats and rankings."""
    monthly_stats: List["TrendsMonthlyStatsResponse"] = []
    latest_ranking: Optional["TrendsPositionRankingResponse"] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================
# Monthly Stats Schemas
# ============================================

class TrendsMonthlyStatsBase(BaseModel):
    """Base schema for monthly stats."""
    location_id: int
    category: str = Field(..., pattern="^(buying|renting)$")
    stats_date: date
    month_year: Optional[str] = None
    view_count: int = 0
    search_percentage: float = 0.0


class TrendsMonthlyStatsResponse(TrendsMonthlyStatsBase):
    """Response schema for monthly stats."""
    id: int
    created_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================
# Position Ranking Schemas
# ============================================

class TrendsPositionRankingBase(BaseModel):
    """Base schema for position rankings."""
    location_id: int
    city_id: int
    category: str = Field(..., pattern="^(buying|renting)$")
    stats_date: date
    current_position: Optional[int] = None
    previous_position: Optional[int] = None
    position_change: Optional[int] = None
    current_search_percentage: Optional[float] = None
    previous_search_percentage: Optional[float] = None
    search_percentage_change: Optional[float] = None
    current_view_count: Optional[int] = None


class TrendsPositionRankingResponse(TrendsPositionRankingBase):
    """Response schema for position rankings."""
    id: int
    location: Optional[TrendsLocationBase] = None
    created_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================
# Metadata Schemas
# ============================================

class TrendsMetadataResponse(BaseModel):
    """Response schema for trends metadata."""
    last_import: Optional[str] = None
    data_version: Optional[str] = None
    buying_total_locations: Optional[int] = None
    renting_total_locations: Optional[int] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================
# Summary/Aggregate Schemas
# ============================================

class TrendsSummaryResponse(BaseModel):
    """Summary response for trends overview."""
    total_regions: int
    total_cities: int
    total_locations: int
    buying_locations: int
    renting_locations: int
    latest_data_date: Optional[date] = None
    last_import: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================
# Search/Filter Params
# ============================================

class TrendsSearchParams(BaseModel):
    """Parameters for searching trends data."""
    category: Optional[str] = Field(None, pattern="^(buying|renting)$")
    region_id: Optional[int] = None
    city_id: Optional[int] = None
    location_id: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    min_position: Optional[int] = None
    max_position: Optional[int] = None
    limit: int = Field(default=50, ge=1, le=500)
    offset: int = Field(default=0, ge=0)

File: frontend/types/api/trends.ts
/**
 * Types for Trends API responses
 */

// ============================================
// Region Types
// ============================================

export interface TrendsRegion {
  id: number;
  name: string;
  name_urdu?: string | null;
  level: number;
  created_at?: string | null;
  updated_at?: string | null;
}

// ============================================
// City Types
// ============================================

export interface TrendsCity {
  id: number;
  name: string;
  name_urdu?: string | null;
  region_id?: number | null;
  latitude?: number | null;
  longitude?: number | null;
  level: number;
  region?: TrendsRegion | null;
  created_at?: string | null;
  updated_at?: string | null;
}

// ============================================
// Location Types
// ============================================

export interface TrendsLocation {
  id: number;
  title: string;
  title_urdu?: string | null;
  city_id?: number | null;
  latitude?: number | null;
  longitude?: number | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface TrendsLocationWithStats extends TrendsLocation {
  monthly_stats: TrendsMonthlyStats[];
  latest_ranking?: TrendsPositionRanking | null;
}

// ============================================
// Monthly Stats Types
// ============================================

export interface TrendsMonthlyStats {
  id: number;
  location_id: number;
  category: "buying" | "renting";
  stats_date: string;
  month_year?: string | null;
  view_count: number;
  search_percentage: number;
  created_at?: string | null;
}

// ============================================
// Position Ranking Types
// ============================================

export interface TrendsPositionRanking {
  id: number;
  location_id: number;
  city_id: number;
  category: "buying" | "renting";
  stats_date: string;
  current_position?: number | null;
  previous_position?: number | null;
  position_change?: number | null;
  current_search_percentage?: number | null;
  previous_search_percentage?: number | null;
  search_percentage_change?: number | null;
  current_view_count?: number | null;
  location?: TrendsLocation | null;
  created_at?: string | null;
}

// ============================================
// Summary Types
// ============================================

export interface TrendsSummary {
  total_regions: number;
  total_cities: number;
  total_locations: number;
  buying_locations: number;
  renting_locations: number;
  latest_data_date?: string | null;
  last_import?: string | null;
}

// ============================================
// Top Location Types (for city rankings)
// ============================================

export interface TopLocation {
  location_id: number;
  title: string;
  title_urdu?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  current_position: number;
  previous_position?: number | null;
  position_change?: number | null;
  current_search_percentage?: number | null;
  current_view_count?: number | null;
  stats_date: string;
}

// ============================================
// Top Mover Types
// ============================================

export interface TopMover {
  location_id: number;
  location_title: string;
  city_id: number;
  city_name: string;
  current_position: number;
  previous_position?: number | null;
  position_change: number;
  current_search_percentage?: number | null;
  current_view_count?: number | null;
}

// ============================================
// Location History Types
// ============================================

export interface LocationHistory {
  location_id: number;
  category: "buying" | "renting";
  labels: string[];
  view_counts: number[];
  search_percentages: number[];
  dates: string[];
}

// ============================================
// Analytics Types
// ============================================

export interface AnalyticsOverview {
  category: "buying" | "renting";
  latest_date?: string | null;
  total_views: number;
  average_search_percentage: number;
  top_cities: {
    id: number;
    name: string;
    total_views: number;
  }[];
  region_distribution: {
    id: number;
    name: string;
    city_count: number;
    location_count: number;
  }[];
}

export interface LocationComparison {
  location_id: number;
  title: string;
  data: {
    labels: string[];
    view_counts: number[];
    search_percentages: number[];
  };
}

// ============================================
// Map Types
// ============================================

export interface MapLocation {
  id: number;
  title: string;
  city_id: number;
  latitude: number;
  longitude: number;
  position?: number | null;
  position_change?: number | null;
  view_count?: number | null;
  search_percentage?: number | null;
}

export interface MapCity {
  id: number;
  name: string;
  region_id?: number | null;
  latitude: number;
  longitude: number;
}

// ============================================
// Filter/Search Params
// ============================================

export interface TrendsSearchParams {
  category?: "buying" | "renting";
  region_id?: number;
  city_id?: number;
  location_id?: number;
  start_date?: string;
  end_date?: string;
  min_position?: number;
  max_position?: number;
  limit?: number;
  offset?: number;
}

export type TrendsCategory = "buying" | "renting";


